Add to explainer: creative scanning via BYOS/V1 trusted scoring signals. (#1406)
WICG/turtledove: FLEDGE.md

  'userBiddingSignals': {...},
  'ads': [{renderUrl: shoesAd1, sizeGroup: 'group1', ...},
          {renderUrl: shoesAd2, sizeGroup: 'group2',

>            creativeScanningMetadata: 'shoemanufacturer.com',

Reviewer: maybe not for this PR, but we should add an example `metadata` here

Reviewer 2: When the TechLab's PA subgroup tackles a creative scanning metadata signaling spec (json) we could share an example from a doc we would publish.

Author: Ack. I've taken a note to create a separate explainer PR for this comment and the one below. Thanks.

>  * `metadata`: Arbitrary metadata that can be used at bidding time.

Reviewer: maybe not for this PR, but we should say what the format is of metadata and userBiddingSignals is in their bullet entries (i.e. "JSON-serializable values (i.e. supported by JSON.stringify()). See [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify)"

Author: Same as above.

>  * `creativeScanningMetadata`: A USVString, no character limit. If `sendCreativeScanningMetadata` is set to true in the auction config, then this string -- along with other metadata as described in [2.1 Initiating an On-Device Auction](#21-initiating-an-on-device-auction) -- is sent as part of the scoring signals fetch request. This provides a way for buyers to convey additional information that sellers need in order to scan an advertisement, such as the associated advertiser domain, which isn't captured in the `renderURL` and other fields sent on the scoring signals fetch request.

Reviewer: would you mind line wrapping this (and any other lines this PR adds or modifies)?
can we linkify USVString like we do above on line 287?
true->`true`
Can we elaborate slightly on "other metadata"?  I don't know what that is.
"and other fields sent on the scoring signals request" is a little confusing (I thought it was attached to the first half of the sentence).  Maybe it might be clearer to write "...which isn't captured in other fields sent on the scoring signals fetch request (e.g. `renderURL`)."

and `adComponents` fields as well as `prevWins`). Note that `include-full-ads` is not compatible
with the auction server, so this mode is only for debugging.

> All fields that accept arbitrary metadata (`userBiddingSignals` and `metadata` field of ads) must be JSON-serializable values (i.e. supported by JSON.stringify()). See [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify). Note that `creativeScanningMetadata` -- unlike `metadata` -- is strictly string-valued, and need not be JSON-serializable.

Reviewer 3: > and need not be JSON-serializable

In practice, though, this will not likely be a scalar, and multiple key- values, like `metadata` will need to be captured here -- does this mean implementers will need to run `JSON.parse()`?

Reviewer 4: We think that us as buyer may need to pass different creative scanning data depending on the seller. SSPs may have different creative validation processes.

Would you consider having `creativeScanningMetadata` as a JSON-serializable object, where keys would be seller domains? Then, Chrome could maybe pass to a given seller's KV server only the `creativeScanningMetadata` relevant to them, i.e. the value corresponding to their domain as key.

There could be a case for this value also being a JSON-serializable object, but having it as a string will be fine for most of our usecases.

Reviewer 5: Executive

`maxTrustedScoringSignalsURLLength` and `trustedScoringSignalsCoordinator` have the same functionality as `maxTrustedBiddingSignalsURLLength` and `trustedBiddingSignalsCoordinator`, as described in [1.2 Interest Group Attributes](#12-interest-group-attributes), but affect the seller's trusted scoring signals fetch as opposed to the trusted bidding signals fetch.

> When set to true, `sendCreativeScanningMetadata` instructs the browser to send a collection of metadata associated with each ad and component ad as part of the scoring signals fetch request, as described in [3.1.1.2 Trusted Scoring Signals](#3112-trusted-scoring-signals).

Reviewer: ```suggestion
When set to `true`, `sendCreativeScanningMetadata` instructs the browser to send the collection of metadata 
stored with each ad and component ad in the `creativeScanningMetadata` field as part of the scoring signals fetch 
request, as described in [3.1.1.2 Trusted Scoring Signals](#3112-trusted-scoring-signals).
```

Author: Accepted your suggestion.

*   bid: A numerical bid value.
*   auctionConfig: The auction configuration object passed to `navigator.runAdAuction()`.
*   trustedScoringSignals: A value retrieved from a real-time trusted server chosen by the seller and reflecting the seller's opinion of this particular creative, as further described in [3.1 Fetching Real-Time Data from a Trusted Server](#31-fetching-real-time-data-from-a-trusted-server) below. This is used when the server is same-origin to the seller; crossOriginTrustedSignals is used otherwise.

> *   browserSignals: An object constructed by the browser, containing information that the browser knows and which the seller's auction script might want to verify. Note that each element in `adComponentsCreativeScanningMetadata` might be either the string value of that component ad's `creativeScanningMetadata`, or null if that component ad didn't provide one.

Reviewer: ```suggestion
*   browserSignals: An object constructed by the browser, containing information that the browser
knows and which the seller's auction script might want to verify.
`adComponentsCreativeScanningMetadata` is a list where each element corresponds to the 
respective element in the `adComponents` list.  Each element might be either the string value
of that component ad's `creativeScanningMetadata`, or `null` if that component ad didn't provide one.
```

Author: Accepted your suggestion.

The `updateIfOlderThanMs` optional field specifies that the interest group should be updated via the `updateURL` mechanism (see the [interest group attributes](#12-interest-group-attributes) section) if the interest group hasn't been joined or updated in a duration of time exceeding `updateIfOlderThanMs` milliseconds. Updates that ended in failure, either parse or network failure, are not considered to increment the last update or join time. An `updateIfOlderThanMs` that's less than 10 minutes will be clamped to 10 minutes.

###### 3.1.1.2 Trusted Scoring Signals

> Similarly, sellers may want to fetch information about a specific creative, e.g. the results of some out-of-band ad scanning system. This works in much the same way as [`trustedBiddingSignalsURL`](#311-trusted-signals-server-with-byos-model). If `trustedScoringSignalsCoordinator` is not present in the auction config, it will send the scoring signals fetch request to a BYOS server. The request base URL is set from the `trustedScoringSignalsURL` property of the seller's auction configuration object. The parameter `experimentGroupId` comes from `sellerExperimentGroupId` in the auction configuration if provided. For bids offered in the auction, the `renderURL` and metadata from both the main ad and component ads of those bids will be appended to the URL using several query parameters.

Reviewer: Perhaps not for this PR, but it's kinda non sequitur how it starts talking about experimentGroupId in the middle of the intro paragraph.  I wonder if we should pull that into a bullet point so that it's easier for folks find definitions of each parameter (without having to look half way through the paragraph).  I also feel like we should start with an example request URL and then go into detailed explanations of each parameter.

Author: I took your suggestion, started with an example, and then changed to bullet points for each query parameter, plus one bullet point above the per-ad/per-component-ad points that explained how those worked wrt parallel lists.
